snapcraft/_store.py

```diff
@@ -50,6 +50,59 @@ def _get_data_from_snap_file(snap_path):
     return snap_yaml
 
 
+def _fail_login(msg=''):
+    logger.info(msg)
+    logger.info('Login failed.')
+    return False
+
+
+def _get_url_from_error(error):
+    if error.extra:
+        return error.extra[0].get('url')
+    return None
+
+
+def _check_dev_agreement_and_namespace_statuses(store):
+    """ Check the agreement and namespace statuses of the dev.
+    Fail if either of those conditions is not met.
+    Re-raise `StoreAccountInformationError` if we get an error and
+    the error is not either of these.
+    """
+    # Check account information for the `developer agreement` status.
+    try:
+        store.get_account_information()
+    except storeapi.errors.StoreAccountInformationError as e:
+        if storeapi.constants.MISSING_AGREEMENT == e.error:
+            # A precaution if store does not return new style error.
+            url = (_get_url_from_error(e) or
+                   storeapi.constants.UBUNTU_STORE_TOS_URL)
+            choice = input(
+                storeapi.constants.AGREEMENT_INPUT_MSG.format(url))
+            if choice == 'y':
+                try:
+                    store.sign_developer_agreement(True)
+                except:
+                    return _fail_login(
+                        storeapi.constants.AGREEMENT_SIGN_ERROR.format(
+                            url))
+            else:
+                return _fail_login(storeapi.constants.AGREEMENT_ERROR)
+
+    # Now check account information for the `namespace` status.
+    try:
+        store.get_account_information()
+    except storeapi.errors.StoreAccountInformationError as e:
+        if storeapi.constants.MISSING_NAMESPACE in e.error:
+            # A precaution if store does not return new style error.
+            url = (_get_url_from_error(e) or
+                   storeapi.constants.UBUNTU_STORE_ACCOUNT_URL)
+            return _fail_login(
+                storeapi.constants.NAMESPACE_ERROR.format(url))
+        else:
+            raise
+    return True
+
+
 def _login(store, acls=None, save=True):
```

Comment:
I don't mind if this just raises an exception that is friendly enough and this `return _login(store)` is refactored a bit.

This may not be a task for you and you may not want to do it, but we have a lot of double logic with exceptions and if True/False around.